README
NAME
    Parse::DMIDecode - Interface to SMBIOS under Linux using dmidecode

SYNOPSIS
     use strict;
     use Parse::DMIDecode ();
     
 my $decoder = new Parse::DMIDecode;
     $decoder->probe; # Actively probe using dmidecode
      
 # Manually supply your own dmidecode output to be parsed
     # $decoder->parse(qx(sudo /usr/sbin/dmidecode));
      
 printf("System: %s, %s",
             $decoder->keyword("system-manufacturer"),
             $decoder->keyword("system-product-name"),
         );

DESCRIPTION
    This module provides an OO interface to SMBIOS information through the
    *dmidecode* command which is known to work under a number of Linux, BSD
    and BeOS variants.

    This module is still actively under development, is no not yet feature
    complete and still needs to be fully documented. There is a possibility
    accessor method names may change before the final release.

METHODS
  new
     my $decoder = Parse::DMIDecode->new(
                         dmidecode => "/usr/sbin/dmidecode"),
                         nowarnings => 1
                     );

  probe
     $decoder->probe;

  parse
     my $raw = qx(sudo /usr/sbin/dmidecode);
     $decoder->prase($raw);

  keyword
     my $serial_number = $decoder->keyword("system-serial-number");

  keywords
     my @keywords = $decoder->keywords;
     my @bios_keywords = $decoder->keywords("bios");
     
 for my $keyword (@bios_keywords) {
         printf("%s => %s\n",
                 $keyword,
                 $decoder->keyword($keyword)
             );
     }

  handle_addresses
     my @addresses = $decoder->handle_addresses;

  get_handles
     use Parse::DMIDecode::Constants qw(@TYPES);
     
 # Available groups to query: bios, system, baseboard,
     #    chassis, processor, memory, cache, connector, slot
     for my $handle ($decoder->get_handles( group => "memory" )) {
         printf(">> Found handle at %s (%s):\n%s\n",
                 $handle->address,
                 $TYPES[$handle->dmitype],
                 $handle->raw
             );
     }

  smbios_version
     my $smbios_version = $decoder->smbios_version;

  dmidecode_version
     my $dmidecode_version = $decoder->dmidecode_version;

    Returns the version number of the copy of *dmidecode* that was used to
    create the source data that was parsed. This value may not be available
    when using older versions of *dmidecode*.

  table_location
     my $memory_address = $decoder->table_location;

  structures
     my $total_structures = $decoder->structures;

SEE ALSO
    examples/*.pl, <http://www.nongnu.org/dmidecode/>,
    <http://linux.dell.com/libsmbios/>,
    <http://sourceforge.net/projects/x86info/>,
    <http://www.dmtf.org/standards/smbios>, biosdecode(8), dmidecode(8),
    vpddecode(8)

VERSION
    $Id: DMIDecode.pm 803 2006-10-22 11:36:22Z nicolaw $

AUTHOR
    Nicola Worthington <nicolaw@cpan.org>

    <http://perlgirl.org.uk>

    If you like this software, why not show your appreciation by sending the
    author something nice from her Amazon wishlist? (
    http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

COPYRIGHT
    Copyright 2006 Nicola Worthington.

    This software is licensed under The Apache Software License, Version
    2.0.

    <http://www.apache.org/licenses/LICENSE-2.0>



Changes
Version 0.02 - 22th October 2006
    + Added better support for older versions of dmidecode
      binary output parsing.
    + Improved examples/test.pl.
    + Added examples/synopsis2.pl from the synopsis of
      Parse::DMIDecode::Handle.
    + Allow get_handles() method to be called without any
      qualifying paramaters so to return *All* handle
      objects.
    + Added description() method to handle objects.
    + Added parsed_structures() method to handle objects.
    + Improved parsing inside handle objects to return
      multiple handle objects where appropriate to allow
      easier access to unique data pairs.

Version 0.01 - 19th October 2006
    + Pre-release.
    + Created Parse::DMIDecode::Handle objects for each
      handle structure.
    + Encapsulated the internal data structure which is
      quite fragile so that people cannot just grab items
      out of it.
    + Added accessor methods for header information.
    + Added selection methods to return handle objects.
    + Added a method to return a list of all known
      keyword strings.
    + Added unit tests.
    + Added a whole bunch more since the first release
      was mostly a placeholder.

Version 0.00 - 17th September 2006
    + Initial developer pre-release to save the namespace
      and to discuss API prior to full release.



examples/test.pl
#!/usr/bin/perl -w

BEGIN { chdir '../' if -d '../examples/'; }

use strict;
use lib qw(./lib);
use Parse::DMIDecode;
use Parse::DMIDecode::Constants qw(@TYPES);

my $decoder = new Parse::DMIDecode;
#$decoder->probe;
#$decoder->parse(qx(sudo /usr/sbin/dmidecode));
$decoder->parse(`cat t/dmidecode_example4.txt`);

#print join("\n",$decoder->keywords)."\n";
#print $decoder->keyword('system-manufacturer')."\n";
#exit;

for my $handle ($decoder->get_handles( group => 'system' )) {
	printf(">>> Found handle at %s (%s):\n >> Keywords: %s\n%s\n",
			$handle->address,
			$TYPES[$handle->dmitype],
			join(', ',$handle->keywords),
			$handle->raw
		);
	for my $keyword ($handle->keywords) {
		printf("  > Keyword %s => '%s'\n",
				$keyword,
				$handle->keyword($keyword)
			);
	}
	print "\n\n";
}






Build.PL
# vim:ts=4:sw=4:tw=78
# $Id: Build.PL 807 2006-10-22 12:15:10Z nicolaw $

use strict;
use Module::Build;
use vars qw($build);

$build = Module::Build->new(
		module_name        => 'Parse::DMIDecode',
		license            => 'open_source',
		create_makefile_pl => 'passthrough',
		create_readme      => 1,
		create_packlist    => 1,
		sign               => 0,

		requires => {
			'Carp'         => 0, # 1.04
			'Scalar::Util' => 0,
			#'Storable'     => 0,
		},

		build_requires => {
			'Test'         => 0, # 1.25
			'Test::More'   => 0, # 0.60
		},

		recommends => {
			'File::Which'         => 0.05, # 0.05
			'Test::Pod'           => 1.20,
			'Test::Pod::Coverage' => 1.06,
			'Test::Deep'          => 0.093,
		},
	);

$build->create_build_script;


# Send perl and module version information home if we've been given
# permission to do so by a human being - default to not send for automated
# testing environments, of if the user does not respond within 20 seconds.

my $url = $ENV{AUTOMATED_TESTING} ? undef : may_send_version_information();
if ($url) {
	my @resp = ();
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 10;
		my $ua = LWP::UserAgent->new(
				agent => 'Build.PL $Revision: 807 $',
				timeout => 9,
				max_size => 500,
			);
		$ua->env_proxy;
		my $response = $ua->get($url);
		if ($response->is_success()) {
			for (split(/\s*\n+\s*/, $response->content())) {
				push @resp, $_ if $_;
			}
		}
		alarm 0;
	};
	print substr($resp[0],0,79) || "Thank you for sending this information.";
	print "\n\n";
}

sub may_send_version_information {
	eval {
		require Config;
		require LWP::UserAgent;
	};
	return undef if $@;

	my $str = sprintf('%s?%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s',
			'http://perlgirl.org.uk/lib/usage.cgi',
			'name',     $build->dist_name(),
			'version',  $build->dist_version(),
			'osname',   $Config::Config{osname},
			'archname', $Config::Config{archname},
			'osver',    $^O,
			'perlver',  $]
		);

	print "\nThank you for downloading ".$build->dist_name()."\n\n";
	print "I would like to find out how many people are using this software,\n";
	print "and on what operating systems and Perl versions. If you have an\n";
	print "internet connection, may I transmit the following information:\n\n";
	print "$str\n\n";

	my $send = 0;
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 20;
		$send = $build->y_n('Send this anonymous information?','n');
		alarm 0;
	};

	return $send ? $str : undef;
}

1;



lib/Parse/DMIDecode.pm
############################################################
#
#   $Id: DMIDecode.pm 803 2006-10-22 11:36:22Z nicolaw $
#   Parse::DMIDecode - Interface to SMBIOS under Linux using dmidecode
#
#   Copyright 2006 Nicola Worthington
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
############################################################

package Parse::DMIDecode;
# vim:ts=4:sw=4:tw=78

use strict;
use Scalar::Util qw(refaddr);
use Parse::DMIDecode::Handle;
use Parse::DMIDecode::Constants qw(@TYPES %GROUPS);
use Carp qw(croak cluck carp);
use vars qw($VERSION $DEBUG);

$VERSION = '0.02' || sprintf('%d', q$Revision: 803 $ =~ /(\d+)/g);
$DEBUG ||= $ENV{DEBUG} ? 1 : 0;

my $objstore = {};


#
# Methods
#

sub new {
	ref(my $class = shift) && croak 'Class name required';
	croak 'Odd number of elements passed when even was expected' if @_ % 2;

	my $self = bless \(my $dummy), $class;
	$objstore->{refaddr($self)} = {@_};
	my $stor = $objstore->{refaddr($self)};

	$stor->{commands} = [qw(dmidecode)];
	my $validkeys = join('|','nowarnings',@{$stor->{commands}});
	my @invalidkeys = grep(!/^$validkeys$/,grep($_ ne 'commands',keys %{$stor}));
	delete $stor->{$_} for @invalidkeys;
	cluck('Unrecognised parameters passed: '.join(', ',@invalidkeys))
		if @invalidkeys && $^W;

	for my $command (@{$stor->{commands}}) {
		croak "Command $command '$stor->{$command}'; file not found"
			if defined $stor->{$command} && !-f $stor->{$command};
	}

	DUMP('$self',$self);
	DUMP('$stor',$stor);
	return $self;
}


sub probe {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);

	my $stor = $objstore->{refaddr($self)};
	eval {
		if (!defined $stor->{dmidecode}) {
			require File::Which;
			for my $command (@{$stor->{commands}}) {
				$stor->{$command} = File::Which::which($command)
					if !defined $stor->{$command};
			}
		}
	};
	croak $@ if $@;

	my ($cmd) = $stor->{dmidecode} =~ /^([\/\.\_\-a-zA-Z0-9 ]+)$/;
	TRACE($cmd);
	croak "dmidecode command '$cmd' does not exist; bum!" if !-f $cmd;

	my $fh;
	local %ENV = %ENV;
	delete @ENV{qw(IFS CDPATH ENV BASH_ENV PATH)};
	open($fh,'-|',$cmd) || croak "Unable to open file handle for command '$cmd': $!";
	while (local $_ = <$fh>) {
		$stor->{raw} .= $_;
	}
	close($fh) || carp "Unable to close file handle for command '$cmd': $!";

	return $self->parse($stor->{raw});
}


sub parse {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);

	my $stor = $objstore->{refaddr($self)};
	my %data = (handles => []);

	my @lines;
	for (@_) {
		push @lines, split(/\n/,$_);
	}

	my $i = 0;
	for (; $i < @lines; $i++) {
		local $_ = $lines[$i];
		if (/^Handle [0-9A-Fx]+/) {
			last;
		} elsif (/^SYSID present\.\s*/) {
			# No-op
		} elsif (/^# dmidecode ([\d\.]+)\s*$/) {
			$data{dmidecode} = $1;
		} elsif (/^(\d+) structures occupying (\d+) bytes?\.\s*$/) {
			$data{structures} = $1;
			$data{bytes} = $2;
		} elsif (/^DMI ([\d\.]+) present\.?\s*$/) {
			$data{dmi} = $1;
		} elsif (/^SMBIOS ([\d\.]+) present\.?\s*$/) {
			$data{smbios} = $1;
		} elsif (/^(?:DMI )?[Tt]able at ([0-9A-Fx]+)\.?\s*$/) {
			$data{location} = $1;
		}
	}

	for (qw(dmidecode structures bytes dmi smbios location)) {
		$data{$_} = undef if !exists $data{$_};
	}

	my $raw_handle_data = '';
	for (; $i < @lines; $i++) {
		if ($lines[$i] =~ /^Handle [0-9A-Fx]+/) {
			push @{$data{handles}}, Parse::DMIDecode::Handle->new(
					raw => $raw_handle_data,
					nowarnings => $stor->{nowarnings}
				) if $raw_handle_data;
			$raw_handle_data = "$lines[$i]\n";
		} else {
			$raw_handle_data .= "$lines[$i]\n";
		}
	}

	push @{$data{handles}}, Parse::DMIDecode::Handle->new(
			raw => $raw_handle_data,
			nowarnings => $stor->{nowarnings}
		) if $raw_handle_data;

	carp sprintf("Only parsed %d structures when %d were expected",
			scalar(@{$data{handles}}), $data{structures}
		) if scalar(@{$data{handles}}) < $data{structures};

	$stor->{parsed} = \%data;
	DUMP('$stor->{parsed}',$stor->{parsed});

	return $stor->{parsed}->{structures};
}


sub get_handles {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);

	croak 'Odd number of elements passed when even was expected' if @_ % 2;
	my %param = @_;
	my $stor = $objstore->{refaddr($self)};
	my @handles;
	my $getall = !keys(%param);

	for my $handle (@{$stor->{parsed}->{handles}}) {
		if ($getall ||
			(defined $param{address} && $handle->address eq $param{address}) ||
			(defined $param{dmitype} && $handle->dmitype == $param{dmitype}) ||
			(defined $param{group} && defined $GROUPS{$param{group}} &&
			 grep($_ == $handle->dmitype,@{$GROUPS{$param{group}}}))
			) {
			push @handles, $handle;
		}
	}

	return @handles;
}


sub structures {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	return $objstore->{refaddr($self)}->{parsed}->{structures};
}


sub table_location {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	return $objstore->{refaddr($self)}->{parsed}->{location};
}


sub smbios_version {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	return $objstore->{refaddr($self)}->{parsed}->{smbios};
}


sub dmidecode_version {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	return $objstore->{refaddr($self)}->{parsed}->{dmidecode};
}


sub handle_addresses {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	return map { $_->handle }
		@{$objstore->{refaddr($self)}->{parsed}->{handles}};
}


sub keywords {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);

	my %keywords;
	my $stor = $objstore->{refaddr($self)};
	for my $handle (@{$stor->{parsed}->{handles}}) {
		for my $keyword ($handle->keywords) {
			$keywords{$keyword} = 1;
		}
	}

	return sort(keys(%keywords));
}


sub keyword {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	croak sprintf('%s elements passed when one was expected',
		(@_ > 1 ? 'Multiple' : 'No')) if @_ != 1;

	my $stor = $objstore->{refaddr($self)};
	for my $handle (@{$stor->{parsed}->{handles}}) {
		if (grep($_ eq $_[0],$handle->keywords)) {
			return $handle->keyword($_[0]);
		}
	}
}


sub DESTROY {
	my $self = shift;
	delete $objstore->{refaddr($self)};
}


sub TRACE {
	return unless $DEBUG;
	carp(shift());
}


sub DUMP {
	return unless $DEBUG;
	eval {
		require Data::Dumper;
		local $Data::Dumper::Indent = 2;
		local $Data::Dumper::Terse = 1;
		carp(shift().': '.Data::Dumper::Dumper(shift()));
	}
}

1;



=pod

=head1 NAME

Parse::DMIDecode - Interface to SMBIOS under Linux using dmidecode

=head1 SYNOPSIS

 use strict;
 use Parse::DMIDecode ();
 
 my $decoder = new Parse::DMIDecode;
 $decoder->probe; # Actively probe using dmidecode
  
 # Manually supply your own dmidecode output to be parsed
 # $decoder->parse(qx(sudo /usr/sbin/dmidecode));
  
 printf("System: %s, %s",
         $decoder->keyword("system-manufacturer"),
         $decoder->keyword("system-product-name"),
     );

=head1 DESCRIPTION

This module provides an OO interface to SMBIOS information through
the I<dmidecode> command which is known to work under a number of
Linux, BSD and BeOS variants.

This module is still actively under development, is no not yet
feature complete and still needs to be fully documented. There is
a possibility accessor method names may change before the final
release.

=head1 METHODS

=head2 new

 my $decoder = Parse::DMIDecode->new(
                     dmidecode => "/usr/sbin/dmidecode"),
                     nowarnings => 1
                 );

=head2 probe

 $decoder->probe;

=head2 parse

 my $raw = qx(sudo /usr/sbin/dmidecode);
 $decoder->prase($raw);

=head2 keyword

 my $serial_number = $decoder->keyword("system-serial-number");

=head2 keywords

 my @keywords = $decoder->keywords;
 my @bios_keywords = $decoder->keywords("bios");
 
 for my $keyword (@bios_keywords) {
     printf("%s => %s\n",
             $keyword,
             $decoder->keyword($keyword)
         );
 }

=head2 handle_addresses

 my @addresses = $decoder->handle_addresses;

=head2 get_handles

 use Parse::DMIDecode::Constants qw(@TYPES);
 
 # Available groups to query: bios, system, baseboard,
 #    chassis, processor, memory, cache, connector, slot
 for my $handle ($decoder->get_handles( group => "memory" )) {
     printf(">> Found handle at %s (%s):\n%s\n",
             $handle->address,
             $TYPES[$handle->dmitype],
             $handle->raw
         );
 }

=head2 smbios_version

 my $smbios_version = $decoder->smbios_version;

=head2 dmidecode_version

 my $dmidecode_version = $decoder->dmidecode_version;

Returns the version number of the copy of I<dmidecode> that was used
to create the source data that was parsed. This value may not be available
when using older versions of I<dmidecode>.

=head2 table_location

 my $memory_address = $decoder->table_location;

=head2 structures

 my $total_structures = $decoder->structures;

=head1 SEE ALSO

examples/*.pl,
L<http://www.nongnu.org/dmidecode/>,
L<http://linux.dell.com/libsmbios/>,
L<http://sourceforge.net/projects/x86info/>,
L<http://www.dmtf.org/standards/smbios>,
L<biosdecode(8)>, L<dmidecode(8)>, L<vpddecode(8)>

=head1 VERSION

$Id: DMIDecode.pm 803 2006-10-22 11:36:22Z nicolaw $

=head1 AUTHOR

Nicola Worthington <nicolaw@cpan.org>

L<http://perlgirl.org.uk>

If you like this software, why not show your appreciation by sending the
author something nice from her
L<Amazon wishlist|http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority>? 
( http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

=head1 COPYRIGHT

Copyright 2006 Nicola Worthington.

This software is licensed under The Apache Software License, Version 2.0.

L<http://www.apache.org/licenses/LICENSE-2.0>

=cut


__END__





lib/Parse/DMIDecode/Constants.pm
############################################################
#
#   $Id: Constants.pm 795 2006-10-20 15:39:17Z nicolaw $
#   Parse::DMIDecode::Constants - SMBIOS Constants
#
#   Copyright 2006 Nicola Worthington
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
############################################################

package Parse::DMIDecode::Constants;
# vim:ts=4:sw=4:tw=78

use strict;
require Exporter;
use vars qw($VERSION $DEBUG
		@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS
		@TYPES %GROUPS %TYPE2GROUP);

@ISA = qw(Exporter);
@EXPORT = qw();
@EXPORT_OK = qw(@TYPES %GROUPS %TYPE2GROUP);
%EXPORT_TAGS = (all => \@EXPORT_OK);

$VERSION = '0.02' || sprintf('%d', q$Revision: 795 $ =~ /(\d+)/g);
$DEBUG ||= $ENV{DEBUG} ? 1 : 0;

@TYPES = ( # Description                    Index Group(s)
		'BIOS',                             # 0   bios
		'System',                           # 1   system
		'Base Board',                       # 2   baseboard
		'Chassis',                          # 3   chassis
		'Processor',                        # 4   processor
		'Memory Controller',                # 5   memory
		'Memory Module',                    # 6   memory
		'Cache',                            # 7   cache
		'Port Connector',                   # 8   connector
		'System Slots',                     # 9   slot
		'On Board Devices',                 # 10  baseboard
		'OEM Strings',                      # 11
		'System Configuration Options',     # 12  system
		'BIOS Language',                    # 13  bios
		'Group Associations',               # 14
		'System Event Log',                 # 15  system
		'Physical Memory Array',            # 16  memory
		'Memory Device',                    # 17  memory
		'32-bit Memory Error',              # 18
		'Memory Array Mapped Address',      # 19
		'Memory Device Mapped Address',     # 20
		'Built-in Pointing Device',         # 21
		'Portable Battery',                 # 22
		'System Reset',                     # 23  system
		'Hardware Security',                # 24
		'System Power Controls',            # 25
		'Voltage Probe',                    # 26
		'Cooling Device',                   # 27
		'Temperature Probe',                # 28
		'Electrical Current Probe',         # 29
		'Out-of-band Remote Access',        # 30
		'Boot Integrity Services',          # 31
		'System Boot',                      # 32  system
		'64-bit Memory Error',              # 33
		'Management Device',                # 34
		'Management Device Component',      # 35
		'Management Device Threshold Data', # 36 
		'Memory Channel',                   # 37
		'IPMI Device',                      # 38
		'Power Supply'                      # 39
	);

%GROUPS = (
		'bios'      => [ qw(0 13) ],
		'system'    => [ qw(1 12 15 23 32) ],
		'baseboard' => [ qw(2 10) ],
		'chassis'   => [ qw(3) ],
		'processor' => [ qw(4) ],
		'memory'    => [ qw(5 6 16 17) ],
		'cache'     => [ qw(7) ],
		'connector' => [ qw(8) ],
		'slot'      => [ qw(9) ],
	);

%TYPE2GROUP = ();
for my $group (keys %GROUPS) {
	for my $dmitype (@{$GROUPS{$group}}) {
		$TYPE2GROUP{$dmitype} = $group;
	}
}


1;


=pod

=head1 NAME

Parse::DMIDecode::Constants - SMBIOS Constants

=head1 SYNOPSIS

 use strict;
 use Parse::DMIDecode::Constants qw(@TYPES %GROUPS);
 
=head1 DESCRIPTION

This module provides and can export constants relating to the SMBIOS
specification and dmidecode interface command.

=head1 EXPORTS

=head2 @TYPES

=head2 %GROUPS

=head1 SEE ALSO

L<Parse::DMIDecode>

=head1 VERSION

$Id: Constants.pm 795 2006-10-20 15:39:17Z nicolaw $

=head1 AUTHOR

Nicola Worthington <nicolaw@cpan.org>

L<http://perlgirl.org.uk>

If you like this software, why not show your appreciation by sending the
author something nice from her
L<Amazon wishlist|http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority>? 
( http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

=head1 COPYRIGHT

Copyright 2006 Nicola Worthington.

This software is licensed under The Apache Software License, Version 2.0.

L<http://www.apache.org/licenses/LICENSE-2.0>

=cut


__END__



